### PR TITLE
[RLlib] APEX-DQN: Bug fix for torch and add learning test.

### LINKED ISSUE #15646: [rllib] no improvement APEX-DQN
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*
`ray == 1.2.0`
`torch == 1.7.1`
`tensorflow == 2.4.1`
I have trained for 22hours but can't see any improvement of episode reward APEX-DQN on pong environment with tuned example [here](https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/dqn/pong-apex.yaml)
Is there anything that I missed on code below?

### Reproduction (REQUIRED)

```
import ray
import ray.rllib.agents.dqn.apex as apex

ray.init()
config = apex.APEX_DEFAULT_CONFIG.copy()
config["framework"] = "torch"
config["target_network_update_freq"] = 20000
config["num_workers"] = 4
config["num_envs_per_worker"] = 8
config["lr"] = .00005
config["train_batch_size"] = 64
config["gamma"] = 0.99
trainer = apex.ApexTrainer(config=config, env="PongNoFrameskip-v4")

for i in range(1000000):
    result = trainer.train()
    print(result['episode_reward_mean'])

    if i % 100 == 0:

        checkpoint = trainer.save()
        print("checkpoint saved at", checkpoint)

```

Screen Shot 2021-05-05 at 5 39 19 PM

### PR DESCRIPTION
**Pull request by sven1977**
APEX-DQN:
- Bug fix: Serialization for torch tensors to remote (replay) actors problematic. Make sure prio-replay indices are numpy, not torch tensors!
- Add APEX-DQN CI learning test for cartpole.

Closes #15646 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/tuned_examples/dqn/cartpole-apex.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+cartpole-apex-dqn:
```
Comment (richardliaw) [inline rllib/tuned_examples/dqn/cartpole-apex.yaml:1]: should you report the results of this tuned example in the comments at top of file?

File: rllib/agents/dqn/learner_thread.py [hunk 2]
```diff
@@ -54,6 +54,7 @@ def step(self):
                         td_error = info.get(
                             "td_error",
                             info[LEARNER_STATS_KEY].get("td_error"))
+                        replay.policy_batches[pid].set_get_interceptor(None)
```
Comment (richardliaw) [inline rllib/agents/dqn/learner_thread.py:57]: what does this do?

File: rllib/tuned_examples/dqn/cartpole-apex.yaml (hunk 1 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/dqn/cartpole-apex.yaml:1]: ^^

File: rllib/agents/dqn/learner_thread.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/dqn/learner_thread.py:57]: I'll comment. It switches off lazy auto-conversion from numpy (SampleBatch) to torch tensors, which was the main reason for this bug.

File: cpp/src/ray/runtime/task/task_executor.cc [hunk 3]
```diff
@@ -1,9 +1,10 @@
 
 #include "task_executor.h"
 
+#include <ray/api/exec_funcs.h>
```
Comment (richardliaw) [inline cpp/src/ray/runtime/task/task_executor.cc:4]: what's this change for?

File: rllib/tuned_examples/dqn/cartpole-apex.yaml (hunk 1 above)
Comment (richardliaw) [reply rllib/tuned_examples/dqn/cartpole-apex.yaml:1]: bump on this @sven1977 ?

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -2818,15 +2818,16 @@ void CoreWorker::HandleExit(const rpc::ExitRequest &request, rpc::ExitReply *rep
...
+  send_reply_callback(
+      Status::OK(),
+      [this, is_idle]() {
+        // If the worker is idle, we exit.
+        if (is_idle) {
+          Exit(rpc::WorkerExitType::INTENDED_EXIT);
+        }
+      },
+      // We need to kill it regardless if the RPC failed.
+      [this]() { Exit(rpc::WorkerExitType::INTENDED_EXIT); });
```
Comment (richardliaw) [inline src/ray/core_worker/core_worker.cc:2830]: should you revert these cpp changes?

File: cpp/src/ray/runtime/task/task_executor.cc (hunk 3 above)
Comment (sven1977) [reply cpp/src/ray/runtime/task/task_executor.cc:4]: Oh no, something got messed up when merging. :) Will revert.

File: src/ray/core_worker/core_worker.cc (hunk 4 above)
Comment (sven1977) [reply src/ray/core_worker/core_worker.cc:2830]: yes

File: rllib/tuned_examples/dqn/cartpole-apex.yaml (hunk 1 above)
Comment (sven1977) [reply rllib/tuned_examples/dqn/cartpole-apex.yaml:1]: We can do that, but usually - for these simple tasks - the stop criteria define what should happen: Learn in 200k steps to get to 100 reward.

File: cpp/src/ray/runtime/task/task_executor.cc (hunk 3 above)
Comment (sven1977) [reply cpp/src/ray/runtime/task/task_executor.cc]: done

File: src/ray/core_worker/core_worker.cc (hunk 4 above)
Comment (sven1977) [reply src/ray/core_worker/core_worker.cc]: done

**Review by richardliaw — APPROVED** [review]
